ci: Automated Y-Stream Releases (#190)
instructlab/dev-docs: docs/ci/ci-automated-release-workflow.md

# Creating Automated Releases: Design Document

## Motivation & Overview

Presently, the release processes for every library within this `instructlab` GitHub organization is entirely manual:

A maintainer has to:

> 1. Manually create a new release branch -- e.g., `release-0.y.0`,

Reviewer: Is this document/automation only meant to cover the y stream releases?

Author: Yes, y-stream only. Sorry, I should clarify that!

At this point, I don't think it's worth automating z-stream releases. We can certainly consider automating them in the future, but they can be complicated due to the fact sometimes backporting and manual rebasing is needed.

Author: I just updated the document to specify y-stream. Can you take a look?

4. Wait for all pull request CI checks to complete,
5. Manually request two maintainers to approve the pull request, and
6. Manually create a release from the GitHub UI using that new branch

This entire process takes at least 10 minutes of manual work, plus however long it takes for the pull request's checks to complete. (In some repositories, like the core repo, this can take 2+ hours.)

Going forward, we should automate these release processes so that contributors and maintainers can focus more on development work and less on creating actual releases.

## Automation Workflow

### Brief Overview of the Automation

The automation logic described in this dev-doc will be published in the form of an in-house GitHub action called `create-automated-release` and it will therefore be callable from any workflow file. Thus, it is strongly recommended that each repository maintainer creates a `.github/workflows/automated-release.yml` workflow file to call this `create-automated-release` GitHub action from.

### Goals of the Automation

The automated process should be:

1. Configurable so that any library maintainer can configure the automation to meet their specific repository's needs

> 2. Schedulable so that releases are generated according to a specific cadence

Reviewer: No objection to this being an option but it's unclear (to me at least) whether we want to use this.

Author: Sometimes maintainers want to create a new y-stream release automatically every X days. For example, the core repo, `instructlab/instructlab`, has a release cadence of 14 days. 

Realistically though, the scheduling portion of this design utilizes the built-in GitHub workflow trigger conditions and is therefore independent of the design itself. Thus, if a maintainer wants to release on a cadence, they can certainly define a schedule in their workflow file, but if they  don't want to do a schedule, then they can just omit the `schedule` trigger condition.

Scheduled releases are generally important for setting expectations around release cadences, but they are by no means required for every library and may not even be applicable to some in this GitHub organization.

### Configurable Components

As mentioned above, there are configurable components within this release process automation. The diagram in the next section references two configurable components:

#### Trigger Schedule

The trigger schedule defines the day and time (in UTC) when the release process will kick off.  This schedule can be disabled if desired, and maintainers can trigger the release process manually when needed instead.

As mentioned in the brief overview of the automation, this in-house `create-automated-release` action will be callable from any workflow file. Thus, each library maintainer who wants to create a scheduled release should first create a  `.github/workflows/automated-release.yml` workflow file in their repository and define a schedule for it. For example:

```yaml
name: Create Y-Stream Release

> on:

Reviewer 2: Would we want to make this also to be available to run outside of this cron schedule for any reason, by exposing the inputs for the workflow ?

Author: Yes, as mentioned near the bottom of the document, there is an option to add your own trigger conditions, like `workflow_dispatch`. When using `workflow_dispatch`, users can manually kick off a release process.

The trigger conditions is defined at the Git workflow level, but not at the `create-automated-release` level.

  schedule:
    - cron: '30 1 1,15 * *' # Triggers at 1:30am UTC every 2 weeks on the 1st and the 15th day of each month
```

#### Custom List of Dependencies to Cap

With each release, some library maintainers may want to cap the version of certain dependencies within their `requirements.txt` file as well as specify the desired upper cap for each one.

> The list of dependencies to cap should be provided in a file called `automated-release-config.yml`. This configuration file may be expanded in the future to accommodate more configurations as needed. Example:

Reviewer 2: @courtneypacheco Would this `automated-release-config.yml` file be per `y` stream release ? Is there a way we could add a checkpoint to verify in the beginning of the workflow to make sure this file exists and if it does, make sure it exists with the necessary details and format ? We could skip the secondary validation, if the file doesn't exist, which seems to be a possibility based on your workflow diagram.

Reviewer 2: Also would it be a good idea to version this file within the `y` stream release, if things change on us ? In other words, do we want this file to be immutable once the release process kicks off.

# Creating Automated Releases: Design Document

## Motivation & Overview

Presently, the release processes for every library within this `instructlab` GitHub organization is entirely manual:

For example, for a typical y-stream release, a maintainer has to:

> 1. Manually create a new release branch -- e.g., `release-0.y.0`,

Reviewer 3: ```suggestion
1. Manually create a new release branch -- e.g., `release-0.y`
```

Author: Good catch on the typo! Will update!

> 2. Manually create a pull request against `release-0.y.0` to cap the versions on some of the dependencies defined in their library's `requirements.txt` file,

Reviewer 3: ```suggestion
2. Manually create a pull request against the `release-0.y` branch to cap the versions on some of the dependencies defined in their library's `requirements.txt` file
```

Author: Thanks for catching that again will update.

> 2. Manually create a pull request against `release-0.y.0` to cap the versions on some of the dependencies defined in their library's `requirements.txt` file,

Reviewer 4: Where is this requirement to cap coming from? Aren't we asked to [not](https://github.com/instructlab/dev-docs/blob/main/docs/dependency-management.md) cap?

```
Only apply "caps" to dependencies (using <) when that dependency has established a pattern of producing new releases with breaking changes.
```

Without this, steps 3-5 are no longer necessary.

Author: We have been capping certain dependencies in the core repo since October 2024: https://github.com/instructlab/instructlab/pulls?q=is%3Apr+%22deps%3A+cap%22+is%3Aclosed+

To elaborate, we have been capping our own InstructLab library dependencies to ensure that if we create a new release from the core repo, that new release won't automatically consume potential breaking changes from one of our own libraries.

Author: Also, to be clear, the "automatic capping" feature is not required to be used by anybody. Maintainers can ignore the feature if desired. And since the automation described here will create a pull request that someone has to manually approve and review, the dependency cappping changes not be merged automatically. 😃

3. Optionally trigger an E2E test against that pull request,
4. Wait for all pull request CI checks to complete,
5. Manually request two maintainers to approve the pull request, and

> 6. Manually create a release from the GitHub UI using that new branch

Reviewer 3: I feel the commas and "and" aren't really necessary in a numbered list

## Generic Automation Workflow: Major Releases, Minor Releases, and Z-Stream Releases

### Brief Overview of the Automation

The automation logic described in this dev-doc will be published in the form of an in-house GitHub action called `create-automated-release` and it will therefore be callable from any workflow file. Thus, it is strongly recommended that each repository maintainer creates a `.github/workflows/automated-release.yml` workflow file to call this `create-automated-release` GitHub action from.

### Goals of the Automation

The automated process should be:

1. Configurable so that any library maintainer can configure the automation to meet their specific repository's needs
2. Schedulable so that releases are generated according to a specific cadence

Scheduled releases are generally important for setting expectations around release cadences, but they are by no means required for every library and may not even be applicable to some in this GitHub organization.

## Y-Stream Automation Workflow

### Overview of Y-Stream (Minor) Release Automation

> Y-stream (minor) releases have historically been handled differently from Z-stream releases. Z-stream releases oftentimes involve backports for bugfixes and may require manual code rebasing to get those backports merged into the appropriate existing release branch. Therefore, we can think of Y-stream release logic as the "basis" for Z-stream release logic, which takes the Y-stream logic and builds upon it to account for backporting and other desirable actions.

Reviewer 3: ```suggestion
Y-stream (minor) releases have historically been handled differently from Z-stream releases. Z-stream releases often times involve backports for bugfixes and may require manual code rebasing to get those backports merged into the appropriate existing release branch. Therefore, we can think of Y-stream release logic as the "basis" for Z-stream release logic, which takes the Y-stream logic and builds upon it to account for backporting and other desirable actions.
```

### Configurable Components

> As mentioned above, there are configurable components within this release process automation. The diagram in the next section references two configurable components:

Reviewer 3: ```suggestion
As mentioned above, there are configurable components within this release process automation. The diagram in the next section references two configurable components
```

#### Trigger Schedule

The trigger schedule defines the day and time (in UTC) when the release process will kick off.  This schedule can be disabled if desired, and maintainers can trigger the release process manually when needed instead.

As mentioned in the brief overview of the automation, this in-house `create-automated-release` action will be callable from any workflow file. Thus, each library maintainer who wants to create a scheduled release should first create a  `.github/workflows/automated-release.yml` workflow file in their repository and define a schedule for it. For example:

```yaml
name: Create Y-Stream Release

on:
  schedule:
    - cron: '30 1 1,15 * *' # Triggers at 1:30am UTC every 2 weeks on the 1st and the 15th day of each month
```

> #### Custom List of Dependencies to Cap

Reviewer 4: See above. Capping is not something we should automate. It's an exceptional situation, not a matter of course.

With each release, some library maintainers may want to cap the version of certain dependencies within their `requirements.txt` file as well as specify the desired upper cap for each one.

The list of dependencies to cap should be provided in a file called `automated-release-config.yml`. This configuration file may be expanded in the future to accommodate more configurations as needed. Example:

```yaml
y_stream_release:

  dependency_caps:
    enable: true # optional parameter. If set to false, then none of the dependencies in `requirements.txt` will be capped.

>     packages:
>       instructlab-sdg: "+0.1.0"
>       instructlab-eval: "+0.2.0"
>       instructlab-training: "+1.0.0"

Reviewer 3: Suggestion here, lmk what you think
```suggestion
    packages:
      - name: instructlab-sdg
        greater_than: 0.1.0
      - name: instructlab-eval
        greater_than: 0.2.0
      - name: instructlab-training
        greater_than: 1.0.0
```

You could have similar fields `equals` or `less_than` when applicable

```

The keys specify the dependencies to cap. (In this case, we only have three: `instructlab-sdg`, `instructlab-eval`, and `instructlab-training`. The other dependencies in the `requirements.txt` file will be ignored and left untouched.)

The value for each key specifies the cap relative to the current lower bound. For example, let's say the current `requirements.txt` file looks like this:

```bash
instructlab-sdg>0.20.0
instructlab-eval>0.1.0
instructlab-training
```

In this case, the automated logic will create a pull request that updates the `requirements.txt` file like so (ignoring the inline explanations I provided as comments):

```bash
instructlab-sdg>0.20.0,<=0.21.0 # increment by 0.1.0 because of the '+0.1.0' in the `cap-deps.cfg` file

> instructlab-eval>0.1.0,<=0.3.0  # increment by 0.2.0 because of the '+0.2.0' in the `cap-deps.cfg` file

Author: Yes it was. Thanks for catching! Will remove that reference. 👍

instructlab-training            # do nothing because there was no lower bound set.
```

If `dependency_caps` is not defined or `enable` is set to `False` under the `dependency_caps` key, then none of the dependencies defined in `requirements.txt` will be capped. If a dependency to cap is defined in `dependency_caps` and that dependency doesn't exist, then it will be ignored.

### Example Workflow File

Below is an example workflow file used to call this in-house GitHub action:

```yaml
name: Create Y-Stream Release

on:
  # Run every Monday at 1AM UTC
  schedule:
    - cron: '0 7 * * 1 ' 

>   # Allow manual dispatch, too

Reviewer 3: ```suggestion
  # Allow manual dispatch too
```
